packages/camera/camera_platform_interface/lib/src/platform_interface/camera_platform.dart

```diff
@@ -0,0 +1,115 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:plugin_platform_interface/plugin_platform_interface.dart';
+import 'package:camera_platform_interface/src/method_channel/method_channel_camera.dart';
+
+import '../../camera_platform_interface.dart';
+
+/// The interface that implementations of camera must implement.
+///
+/// Platform implementations should extend this class rather than implement it as `camera`
+/// does not consider newly added methods to be breaking changes. Extending this class
+/// (using `extends`) ensures that the subclass will get the default implementation, while
+/// platform implementations that `implements` this interface will be broken by newly added
+/// [CameraPlatform] methods.
+abstract class CameraPlatform extends PlatformInterface {
+  /// Constructs a CameraPlatform.
+  CameraPlatform() : super(token: _token);
+
+  static final Object _token = Object();
+
+  static CameraPlatform _instance = MethodChannelCamera();
+
+  /// The default instance of [CameraPlatform] to use.
+  ///
+  /// Defaults to [MethodChannelCamera].
+  static CameraPlatform get instance => _instance;
+
+  /// Platform-specific plugins should set this with their own platform-specific
+  /// class that extends [CameraPlatform] when they register themselves.
+  static set instance(CameraPlatform instance) {
+    PlatformInterface.verifyToken(instance, _token);
+    _instance = instance;
+  }
+
+  /// Completes with a list of available cameras.
+  Future<List<CameraDescription>> availableCameras() {
+    throw UnimplementedError('availableCameras() is not implemented.');
+  }
+
+  /// Initializes the camera on the device and returns its textureId.
+  Future<int> initializeCamera(
+    CameraDescription cameraDescription, {
+    ResolutionPreset resolutionPreset,
+    bool enableAudio,
+  }) {
+    throw UnimplementedError('initializeCamera() is not implemented.');
+  }
+
+  /// Returns a Stream of [CameraEvent]s.
+  Stream<CameraEvent> cameraEventsFor(int textureId) {
+    throw UnimplementedError('videoEventsFor() has not been implemented.');
+  }
```

Comment:
The [events](https://github.com/flutter/plugins/blob/master/packages/camera/lib/camera.dart#L371-L377) coming from the native side are pretty scarce. Should `CameraImage` events be wrapped through this API too?

Also: this big stream of events may be split into separate Streams, one for each type of event, so the API ends up being a little bit more semantic (maybe even in the frontend of the plugin!).

Instead of making people do:

```dart
final eventSubscription = cameraEventsFor(textureId).listen((event) {
  if (event.type == X) {
    handleX(event);
  } else if (event.type == Y) {
    handleY(event);
  }
});
```

Different streams can be provided (one per EventType), so people can do this instead:

```dart
final eventSubscription = onX(textureId).listen(handleX);
// Elsewhere in the code...
final anotherEventSubscriptionMaybe = onY(textureId).listen(handleY);
```

We've done something similar in Google Maps (it has 11 separate event streams, and growing). [Interface](https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/platform_interface/google_maps_flutter_platform.dart#L249), [method channel](https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart#L65), [web](https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_flutter_web.dart#L26). I used `whereType` to have one big central stream that gets split on sub-streams, but the `where` method is probably enough if you don't want to create a class hierarchy of CameraEvents.